src/openfl/utils/_internal/ShaderMacro.hx

```diff
@@ -199,6 +442,249 @@ class ShaderMacro
+				if (isFragment)
+				{
+					result = varyingKeyword.replace(result, "in $1 $2");
+				}
+				else
+				{
+					result = attributeKeyword.replace(result, "in $1 $2");
+					result = varyingKeyword.replace(result, "out $1 $2");
+				}
+				result = texture2DKeyword.replace(result, "texture");
+				result = glFragColorKeyword.replace(result, "fragColor");
+				return result;
+
+			case "310 es":
+				var result = processGLSLText(source, "300 es", isFragment);
+				return result;
+
+			case "320 es":
+				var result = processGLSLText(source, "310 es", isFragment);
+				return result;
+
+			case "330":
+				#if desktop
+				var result = processGLSLText(source, "320 es", isFragment);
+				#else
+				var result = source;
+				#end
+				return result;
+			case "400":
+				var result = processGLSLText(source, "330", isFragment);
+				return result;
+			case "410":
+				var result = processGLSLText(source, "400", isFragment);
+				return result;
+			case "420":
+				var result = processGLSLText(source, "410", isFragment);
+				return result;
+			case "430":
+				var result = processGLSLText(source, "420", isFragment);
+				return result;
+			case "440":
+				var result = processGLSLText(source, "430", isFragment);
+				return result;
+			case "450":
+				var result = processGLSLText(source, "440", isFragment);
+				return result;
+			case "460":
+				var result = processGLSLText(source, "450", isFragment);
+				return result;
+		}
+	}
+
+	private static function buildGLSLHeaders(glVersion:String):String
+	{
+		if (StringTools.endsWith(glVersion, " compatibility")) return "";
+		if (StringTools.endsWith(glVersion, " core")) return buildGLSLHeaders(StringTools.replace(glVersion, " core", ""));
+
+		switch (glVersion)
+		{
+			default:
+				return "";
+
+			case "100":
+				return "";
+			case "110":
+				return "";
+			case "120":
+				return "";
+			case "130":
+				return "";
+			case "140":
+				return "";
+			case "150":
+				return "";
+
+			case "300 es":
+				#if desktop
+				var result = "layout (location = 0) out vec4 fragColor;\n";
+				#else
+				var result = "out vec4 fragColor;\n";
+				// var result = "";
+				#end
+				return result;
+			case "310 es":
+				var result = buildGLSLHeaders("300 es");
+				return result;
+			case "320 es":
+				var result = buildGLSLHeaders("310 es");
+				return result;
+
+			case "330":
+				var result = buildGLSLHeaders("320 es");
+				return result;
+
+			case "400":
+				var result = buildGLSLHeaders("330");
+				return result;
+			case "410":
+				var result = buildGLSLHeaders("400");
+				return result;
+			case "420":
+				var result = buildGLSLHeaders("410");
+				return result;
+			case "430":
+				var result = buildGLSLHeaders("420");
+				return result;
+			case "440":
+				var result = buildGLSLHeaders("430");
+				return result;
+			case "450":
+				var result = buildGLSLHeaders("440");
+				return result;
+			case "460":
+				var result = buildGLSLHeaders("450");
+				return result;
+		}
```

Comment:
this also seems needlessly overcomplicated, compared to something like:

```suggestion
		
		return switch (glVersion)
		{
			#if desktop
			case "300 es": "layout (location = 0) out vec4 fragColor;\n";
			#else
			case "300 es": "out vec4 fragColor;\n";
			#end
			case "310 es": buildGLSLHeaders("300 es");
			case "320 es": buildGLSLHeaders("310 es");
			case "330": buildGLSLHeaders("320 es");
			case "400": buildGLSLHeaders("330");
			case "410": buildGLSLHeaders("400");
			case "420": buildGLSLHeaders("410");
			case "430": buildGLSLHeaders("420");
			case "440": buildGLSLHeaders("430");
			case "450": buildGLSLHeaders("440");
			case "460": buildGLSLHeaders("450");
			default: "";
		}
```

Reply:
This is nice, I forgot about return switch.